src/ai/backend/manager/idle.py

```diff
@@ -326,6 +378,147 @@ async def update_app_streaming_status(
     ) -> None:
         pass
 
+    async def _do_idle_check(
+        self,
+        context: None,
+        source: AgentId,
+        event: DoIdleCheckEvent,
+    ) -> None:
+        cache_token = self._policy_cache.set(dict())
+        try:
+            return await super()._do_idle_check(context, source, event)
+        finally:
+            self._policy_cache.reset(cache_token)
+
+    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
+        session_id = session["id"]
+        occupied_slots = session["occupied_slots"]
+        interval = self.timer.interval
+        window_size = int(self.time_window.total_seconds() / interval)
+
+        # Respect idle_timeout, from keypair resource policy, over time_window.
+        policy_cache = self._policy_cache.get()
+        policy = policy_cache.get(session["access_key"], None)
+        if policy is None:
+            query = (
+                sa.select([keypair_resource_policies])
+                .select_from(
+                    sa.join(
+                        keypairs,
+                        keypair_resource_policies,
+                        (
+                            keypair_resource_policies.c.name
+                            == keypairs.c.resource_policy
+                        ),
+                    )
+                )
+                .where(keypairs.c.access_key == session["access_key"])
+            )
+            result = await dbconn.execute(query)
+            policy = result.first()
+            assert policy is not None
+            policy_cache[session["access_key"]] = policy
+        if policy["idle_timeout"] >= 0:
+            window_size = int(float(policy["idle_timeout"]) / interval)
+        if (window_size <= 0) or (math.isinf(window_size) and window_size > 0):
+            return True
+
+        # Get current utilization data from all containers of the session.
+        if session["cluster_size"] > 1:
+            query = (
+                sa.select([kernels.c.id])
+                .select_from(kernels)
+                .where(
+                    (kernels.c.session_id == session_id) &
+                    (kernels.c.status.in_(LIVE_STATUS))
+                )
+            )
+            result = await dbconn.execute(query)
+            rows = result.fetchall()
+            kernel_ids = [k["id"] for k in rows]
+        else:
+            kernel_ids = [session_id]
+        current_utilizations = await self.get_current_utilization(kernel_ids, occupied_slots)
+        if current_utilizations is None:
+            return True
+
+        # Update utilization time-series data.
+        not_enough_data = False
+        util_series_key = f"session.{session_id}.util_series"
+        raw_util_series = await self._redis.get(util_series_key)
+        try:
+            util_series = json.loads(raw_util_series)
```

Comment:
I'd suggest to use `msgpack` for faster serialization & deserialization and less memory usage in Redis, unless there are specific reasons to make it human-readable in the memory.

Reply:
msgpack can't unpack this type of stored data with values in the list
{'cpu': [0.42, 0.44], 'cuda_mem': [0.0, 0.0], 'cuda_util': [0.0, 0.0], 'mem': [30.0, 30.0]}

Reply:
Fixed the inability to unpack list values as lists in lablup/backend.ai-common#79 . Now you may pass `use_list=True` to our msgpack wrapper functions to preserve lists for history updates.

Reply:
The problem is resolved by lablup/backend.ai-common#79, but it was not backported to 21.03. So, I backported by
https://github.com/lablup/backend.ai-common/commit/ed4d567f841340f3612e34a215c007d19311f77f, and I think we can use msgpack to pack/unpack the utilization series data.